### PR TITLE
On-OBC temperature sensor driver

### PR DESCRIPTION
**Pull request by parkerg-13**
- Includes tests

Couple things:
1) I'm not too sure how to write unit tests for the read_temperature function since it interacts with the temperature device.
2) There seems to be a weird issue where after I measure a cold temperature it sticks to that value for quite some time and I can't seem to get it to change (warming it up with my thumb for example). I'm not sure why this is happening.
3) We need to have a discussion on the precision of the temperature and whether we want this to be changeable or not. Right now its hardcoded as a const.

### CODE REVIEW

File: firmware/Core/Inc/temperature_sensor/temperature_sensor.h [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    TWO_BIT_PRECISION,
+    THREE_BIT_PRECISION,
+    FOUR_BIT_PRECISION,
+} DataPrecision;
+
+int read_temperature(float *result);
+
+float calculate_decimal_portion(uint8_t bits, DataPrecision precision);
+
+float calculate_integer_portion(uint8_t bits);
```
Comment (NuclearTea) [inline firmware/Core/Inc/temperature_sensor/temperature_sensor.h:18]: is this function implemented, if not remove this declaration

File: firmware/Core/Inc/temperature_sensor/temperature_sensor.h [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+typedef enum  
+{
+    ONE_BIT_PRECISION = 0,
+    TWO_BIT_PRECISION,
+    THREE_BIT_PRECISION,
+    FOUR_BIT_PRECISION,
+} DataPrecision;
+
+int read_temperature(float *result);
```
Comment (NuclearTea) [inline firmware/Core/Inc/temperature_sensor/temperature_sensor.h:14]: I think this should return a uint8_t because the return values are always >=0

**Review by NuclearTea — CHANGES_REQUESTED** [review]
Overall, great work!

Just some small tweaks to naming and some helpful additions to the comments.

Great Job!

File: firmware/Core/Inc/temperature_sensor/temperature_sensor.h [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+#include <stdint.h>
+
+typedef enum  
+{
+    ONE_BIT_PRECISION = 0,
+    TWO_BIT_PRECISION,
+    THREE_BIT_PRECISION,
+    FOUR_BIT_PRECISION,
+} DataPrecision;
```
[the remark this answers is not in the document]
Comment (parkerg-13) [reply firmware/Core/Inc/temperature_sensor/temperature_sensor.h]: Done.

File: firmware/Core/Src/temperature_sensor/temperature_sensor_driver.c [hunk 4]
```diff
@@ -0,0 +1,88 @@
+#include "main.h"
+#include <stdint.h>
+#include "temperature_sensor.h"
```
[the remark this answers is not in the document]
Comment (parkerg-13) [reply firmware/Core/Src/temperature_sensor/temperature_sensor_driver.c:5]: Done.

File: firmware/Core/Src/temperature_sensor/temperature_sensor_driver.c [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+#include "stdio.h"
+
+const uint16_t device_addr =  0x91;
+const temp_register_addr  = 0x00;
+const config_register_addr = 0x01;
+const DataPrecision dataPrecision = ONE_BIT_PRECISION;
```
[the remark this answers is not in the document]
Comment (parkerg-13) [reply firmware/Core/Src/temperature_sensor/temperature_sensor_driver.c]: Done.

File: firmware/Core/Src/temperature_sensor/temperature_sensor_driver.c [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+/// @brief converts the bits provided to a decimal value based on the provided precision. Works on the four MSBs.
+/// @param bits
+/// - Arg 0: Bits to convert to decimal value
+/// @param dataPrecision
+/// - Arg 1: Precision to which we want the output decimal value
+/// @return the decimal after conversion.
+float calculate_decimal_portion(uint8_t bits, DataPrecision dataPrecision)
+{
+    float decimal = 0.0f;
+    uint8_t maskedBits = 0;
```
[the remark this answers is not in the document]
Comment (parkerg-13) [reply firmware/Core/Src/temperature_sensor/temperature_sensor_driver.c]: Done.

File: firmware/Core/Src/temperature_sensor/temperature_sensor_driver.c [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    return 0;
+}
+
+/// @brief converts the bits provided to a decimal value based on the provided precision. Works on the four MSBs.
+/// @param bits
+/// - Arg 0: Bits to convert to decimal value
+/// @param dataPrecision
+/// - Arg 1: Precision to which we want the output decimal value
+/// @return the decimal after conversion.
+float calculate_decimal_portion(uint8_t bits, DataPrecision dataPrecision)
```
[the remark this answers is not in the document]
Comment (parkerg-13) [reply firmware/Core/Src/temperature_sensor/temperature_sensor_driver.c]: Done.

File: firmware/Core/Src/temperature_sensor/temperature_sensor_driver.c [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    // calculate the value to the right of the decimal.
+    float decimalPortion = calculate_decimal_portion(buf[1], dataPrecision);
+
+    // add the integer portion with the decimal portion and store it.
+    *result = (float)((int8_t)buf[0]) + decimalPortion;
+
+    return 0;
+}
+
+/// @brief converts the bits provided to a decimal value based on the provided precision. Works on the four MSBs.
```
[the remark this answers is not in the document]
Comment (parkerg-13) [reply firmware/Core/Src/temperature_sensor/temperature_sensor_driver.c]: Done.

File: firmware/Core/Inc/temperature_sensor/temperature_sensor.h (hunk 1 above)
Comment (parkerg-13) [reply firmware/Core/Inc/temperature_sensor/temperature_sensor.h]: Done.

File: firmware/Core/Inc/temperature_sensor/temperature_sensor.h (hunk 2 above)
Comment (parkerg-13) [reply firmware/Core/Inc/temperature_sensor/temperature_sensor.h]: Changed.

File: firmware/Core/Inc/temperature_sensor/temperature_sensor.h [hunk 9]
```diff
@@ -0,0 +1,18 @@
+#ifndef __INCLUDE_GUARD_TEMPERATURE_SENSOR_H__
+#define __INCLUDE_GUARD_TEMPERATURE_SENSOR_H__
+
+#include <stdint.h>
+
+typedef enum  
+{
+    ONE_BIT_PRECISION = 0,
```
Comment (NuclearTea) [inline firmware/Core/Inc/temperature_sensor/temperature_sensor.h:8]: can you prepend all of these with TEMPERATURE_SENSOR_

**Review by NuclearTea — CHANGES_REQUESTED** [review]
Nice Work!

I think you are failing one of the github actions, please resolve that.

**Comment by parkerg-13** [comment]
I think the action is fixed now.

File: firmware/Core/Inc/temperature_sensor/temperature_sensor.h (hunk 9 above)
Comment (parkerg-13) [reply firmware/Core/Inc/temperature_sensor/temperature_sensor.h]: Fixed.

**Comment by DeflateAwning** [comment]
Please re-architect this so that there is a function which converts the raw 2-byte array to a float (or a signed int, representing the value in "centi-degrees celcius"). Please add unit tests for this function. Please collect the unit tests from Page 15 (Section 2.7) of [the linked datasheet](https://www.st.com/resource/en/datasheet/stds75.pdf), as well as some of your own.

I see an enum with "1 bit precision", "2 bit precision", etc. Not sure what the idea is (maybe they map to the 9-bit, 10-bit, etc.). Would suggest changing.

Calculating the fractional decimal bits is weird. I'd recommend determining the signed value, and then multiplying by the scaling factor (the precision) of the measurement, based on the number of bits.

```c
// The following goes in a function like "float TEMP_SENSOR_convert_raw_to_deg_c(uint8_t buf[]) {}"

uint16_t val = (rx_data[1] << 8) & rx_data[0];

// TODO: handle the sign bit correctly

float temp_C = ((float) val) / (0.5); // TODO: select the 0.5 coefficient based on the number of bits of precision

return temp_C;
```

**Comment by parkerg-13** [comment]
I will re-architect the function and update the naming of the precision bits for more clarity as you suggested tomorrow. I'm not really sure about the temperature calculation though and how it works. Maybe I misunderstood something in the datasheet. I'll have to talk to you about it tomorrow evening.

**Comment by DeflateAwning** [comment]
Is this ready for review? Can you please rebase or merge main into this branch to fix the conflict?

**Comment by DeflateAwning** [comment]
I will review this only once more. Please add unit tests before review.

File: firmware/Core/Inc/temperature_sensor/temperature_sensor.h [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+typedef enum  
+{
+    TEMP_SENSOR_TWELVE_BIT_PRECISION_INSIGNIFICANT_BYTES = 4,
+    TEMP_SENSOR_ELEVEN_BIT_PRECISION_INSIGNIFICANT_BYTES,
+    TEMP_SENSOR_TEN_BIT_PRECISION_INSIGNIFICANT_BYTES,
+    TEMP_SENSOR_NINE_BIT_PRECISION_INSIGNIFICANT_BYTES,
+} Temperature_Sensor_Data_Precision_Insignificant_Bytes_t;
+
+
+uint8_t read_temperature(int32_t *result);
```
Comment (NuclearTea) [inline firmware/Core/Inc/temperature_sensor/temperature_sensor.h:20]: This function should follow the same naming formatting as the others

File: firmware/Core/Src/telecommands/temperature_sensor_telecommand_defs.c [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+/// @return 0 if successful, 1 if error.
+uint8_t TCMDEXEC_read_temperature(const char *args_str, TCMD_TelecommandChannel_enum_t tcmd_channel, 
+                                char *response_output_buf, uint16_t response_output_buf_len)
+{
+
+    int32_t temperature;
+
+    uint8_t success_result = read_temperature(&temperature);
+
+    if (success_result == 0)
```
[the remark this answers is not in the document]
Comment (NuclearTea) [reply firmware/Core/Src/telecommands/temperature_sensor_telecommand_defs.c:33]: This still has not been addressed

File: firmware/Core/Src/temperature_sensor/temperature_sensor_driver.c [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+Temperature_Sensor_Data_Precision_Insignificant_Bytes_t TEMP_SENSOR_precision_insignificant_bits = TEMP_SENSOR_TEN_BIT_PRECISION_INSIGNIFICANT_BYTES;
+
+
+/// @brief Reads the temperature from the STDS75DS2F and stores it in the provided variable pointer result. 
+///         Precision can be configured from 1 to 4 bits.
+///         Temperature range is -55 to 125 degrees celsius with +/- 3 degrees celsius over the whole range.
+/// @param result
+/// - Arg 0: Memory location to store the temperature result.
+/// @return 0 if successful, 1 if config register error, 2 if temp register error.
+uint8_t read_temperature(int32_t *result)
```
Comment (NuclearTea) [inline firmware/Core/Src/temperature_sensor/temperature_sensor_driver.c:34]: Like the comment above, this should follow the naming convention you used for the other functions

**Review by NuclearTea — CHANGES_REQUESTED** [review]
Better. Still some small things remaining

**Comment by parkerg-13** [comment]
@DeflateAwning Its ready for you now

**Review by NuclearTea — COMMENTED** [review]
Everything looks good to me.

This solution is probably a bit overkill for a temperature sensor driver, but I think its approvable. Will default to @DeflateAwning to approve this.

**Comment by DeflateAwning** [comment]
Just a heads up that @aaryandeshpande is now waiting on this to continue OBC testing. Can we please finish this today?

**Comment by parkerg-13** [comment]
@DeflateAwning Has been pushed. I marked where we tested up to in the commit message. There is one more commit after which changes the output string to do C * 10, C * 100, etc as we discussed.

**Comment by DeflateAwning** [comment]
There are changes after the message saying when it was tested. We will test it again just to make sure it's still good!

**Comment by parkerg-13** [comment]
Example outputs for each precision.

Screenshot 2024-10-29 at 7 12 28 PM
Screenshot 2024-10-29 at 7 12 47 PM
Screenshot 2024-10-29 at 7 13 21 PM
Screenshot 2024-10-29 at 7 13 37 PM

**Review by DeflateAwning — APPROVED** [review]
Thanks!

**Review by NuclearTea — APPROVED** [review]
LGTM!

**Pull request merged**